refactor(PXE): clean up block sync

mverzilli opened the pull request:
Block sync is currently handled through collaboration of `Synchronizer`, `SyncDataProvider`, and even `PXEOracleInterface#getAnchorBlock`. 

## Current problems
- The names aren't very intuitive (`Synchronizer` and `SyncDataProvider` only handle chain state data at the end of the day).
- Any code running inside `ContractFunctionSimulator` can call `getAnchorBlock` whenever it pleases. This works currently because calls to `Synchronizer` are carefully orchestrated from PXE. But it seems error prone, obscure, and it gives the false impression that function simulation happens on live block data when it actually depends on a static anchor.
- There are multiple ways throughout the app to get the anchor block: `Synchronizer#getSynchedBlock` (unused), `PXEOracleInterface#getAnchorBlock` (unnecessary, inappropriate), `SyncDataProvider#getBlockNumber` (redundant), and `SyncDataProvider#getBlockHeader` (the one that should be the source of truth)
- Many checks are relying on block number when they should rely on block hash. 

## Refactor 

(note: adjusted naming after discussing with @benesjan) 

- Rename `Synchronizer` => `BlockSynchronizer`
- Rename `SyncDataProvider` => `AnchorBlockDataProvider`
- Remove `AnchorBlockDataProvider#getBlockNumber`. `AnchorBlockDataProvider#getBlockHeader` remains the only way to consume current block information for the rest of the app. To get the block number, there's a simple `getBlockNumber` accessor in `BlockHeader`.
- Remove `ExecutionDataProvider#getAnchorBlock`. Instead, explicitly pass anchor block headers as params to functions in "the realm" of `ContractFunctionSimulator` (`run`, `runUtility`, `varifyCurrentClassId`, etc).   

## Out of scope of PR

> - Many checks are relying on block number when they should rely on block hash. 

This should be tackled on a per use case/feature basis in future PRs, though relying on `getBlockHeader` solely seems aligned in the right direction. 

Closes F-225

benesjan commented on yarn-project/pxe/src/chain_state_synchronizer/chain_state_synchronizer.ts:
```diff
@@ -5,23 +5,23 @@ import { L2BlockStream, type L2BlockStreamEvent, type L2BlockStreamEventHandler
  * for querying sync status.
  */
-export class Synchronizer implements L2BlockStreamEventHandler {
+export class ChainStateSynchronizer implements L2BlockStreamEventHandler {
```
I am not sure I like the renaming because chain state is vague  - chain state can be basically anything from private logs to archiver tree root etc.

So it feels like the new name doesn't give you much new valuable info. Also when I see synchronizer in PXE package I think of PXE synchronizer which makes me think of exactly this - synchronizing blocks from node and handling reorgs etc.

Or am I just brain damaged from looking at this code for too long?

mverzilli replied:
> Or am I just brain damaged from looking at this code for too long?

Maybe a little bit? :grimacing: 

There's many synchronization tasks going at many levels, so `Synchronizer` IMO is indeed too vague. Based on your description, how about `BlockSynchronizer` and `BlockDataProvider` then?

benesjan replied:
I like `BlockSynchronizer` 👍

benesjan commented on yarn-project/pxe/src/chain_state_synchronizer/chain_state_synchronizer.ts:
```diff
@@ -5,23 +5,23 @@ import { L2BlockStream, type L2BlockStreamEvent, type L2BlockStreamEventHandler
   constructor(
     private node: AztecNode,
-    private syncDataProvider: SyncDataProvider,
+    private chainStateDataProvider: ChainStateDataProvider,
```
Given that this provider just stores the anchor block header maybe we could call it `AnchorBlockDataProvider`?

mverzilli replied:
That's a good option! Or just `BlockDataProvider`? A bit shorter, concrete enough?

benesjan replied:
As discussed in DMs given that it stores only the anchor block I think it's nice to `AnchorBlockDataProvider`

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution.test.ts:
```diff
@@ -1153,9 +1171,6 @@ describe('Private Execution test suite', () => {
   describe('Historical header in private context', () => {
```
```suggestion
  describe('Anchor block header in private context', () => {
```

unrelated - the naming here was stale (we use historical header before Mike came with the cute "anchor block" name)

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/oracle_version_is_checked.test.ts:
```diff
@@ -68,7 +67,7 @@ describe('Oracle Version Check test suite', () => {
       const msgSender = await AztecAddress.random();
       const senderForTags = await AztecAddress.random();
-      await acirSimulator.run(txRequest, contractAddress, selector, msgSender, senderForTags);
+      await acirSimulator.run(txRequest, contractAddress, selector, msgSender, BlockHeader.empty(), senderForTags);
```
Unrelated to your PR as this was also undocumented before but I think it would be nice to have a `getGenesisBlockHeader` function and use it everywhere in the tests.

mverzilli replied:
I can look into that, would it be just a wrapper for `BlockHeader.empty()` or should we be testing against something different?

benesjan replied:
I was lying and genesis block header is actually not an empty block header (see [here](https://github.com/AztecProtocol/aztec-packages/blob/9276498bb86449968c9d4d0fda7a218b8ef48e60/noir-projects/noir-protocol-circuits/crates/types/src/abis/block_header.nr#L61-L77)).

Given that there is [this comment](https://github.com/AztecProtocol/aztec-packages/blob/9276498bb86449968c9d4d0fda7a218b8ef48e60/yarn-project/pxe/src/synchronizer/synchronizer.ts#L115) in Synchronizer I think it would make sense to tackle this in a followup PR.

So feel free to ignore my comment for now.

benesjan replied:
Was thinking about this a bit more and I think it might be better to just a random block header here because the tests probably don't care about it at all here. Having it be random would make this explicit.

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/oracle_version_is_checked.test.ts:
```diff
@@ -97,7 +96,7 @@ describe('Oracle Version Check test suite', () => {
 
       // Call the utility function
-      await acirSimulator.runUtility(execRequest, [], []);
+      await acirSimulator.runUtility(execRequest, [], BlockHeader.empty(), []);
```
Also here - `getGenesisBlockHeader`

benesjan commented on yarn-project/pxe/src/contract_function_simulator/oracle/utility_execution.test.ts:
```diff
@@ -83,7 +82,7 @@ describe('Utility Execution test suite', () => {
     };
 
-    const result = await acirSimulator.runUtility(execRequest, [], []);
+    const result = await acirSimulator.runUtility(execRequest, [], BlockHeader.empty(), []);
```
Also here - `getGenesisBlockHeader`

benesjan commented on yarn-project/pxe/src/contract_function_simulator/contract_function_simulator.ts:
```diff
@@ -108,13 +109,13 @@ export class ContractFunctionSimulator {
     contractAddress: AztecAddress,
     selector: FunctionSelector,
     msgSender = AztecAddress.fromField(Fr.MAX_FIELD_VALUE),
+    anchorBlockHeader: BlockHeader,
```
Stale function docs

benesjan commented on yarn-project/pxe/src/contract_function_simulator/contract_function_simulator.ts:
```diff
@@ -217,16 +218,29 @@ export class ContractFunctionSimulator {
    */
-  public async runUtility(call: FunctionCall, authwits: AuthWitness[], scopes?: AztecAddress[]): Promise<Fr[]> {
-    await verifyCurrentClassId(call.to, this.executionDataProvider);
+  public async runUtility(
```
Stale function docs - the docs might be too repetitive here though as I think it might not be necessary to have the param descriptions everywhere. But that's for another PR if we decide to do that.

benesjan commented on yarn-project/pxe/src/storage/chain_state_data_provider/chain_state_data_provider.ts:
```diff
@@ -1,8 +1,7 @@
 import { BlockHeader } from '@aztec/stdlib/tx';
 
-export class SyncDataProvider {
+export class ChainStateDataProvider {
```
As mentioned before would rename this to `AnchorBlockDataProvider`.

With the new name I think it would be fine to have the specific getters like `get number()` or `get hash()`. 

Looking through this PR I think we don't care about the full header in the majority of the places so having these getters would make it all cleaner.

(Random thought: Would it work to just pass around the anchor block hash through the simulators or do we really need the other info in the header?)

mverzilli replied:
Yeah I had that thought too, but it wasn't the point of the PR so I think relying on the header for now is a good intermediate step (and a bit better than bare block numbers)

benesjan reviewed (commented):
This is fantastic and makes a ton of sense.

Just not sure about some of the naming. Will be happy to approve once we discuss that.

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution.test.ts:
```diff
@@ -647,13 +653,16 @@ describe('Private Execution test suite', () => {
           runSimulator({
             contractAddress,
             artifact: TestContractArtifact,
+            anchorBlockHeader: previousAnchorBlock,
```
Very nice

nventuro commented on yarn-project/stdlib/src/block/l2_block_stream/l2_block_stream.ts:
```diff
@@ -26,6 +26,10 @@ export class L2BlockStream {
       skipFinalized?: boolean;
     } = {},
   ) {
+    // Note that RunningPromise is in stopped state by default. This promise won't run until someone invokes `start`,
+    // which makes it run periodically, or `sync`, which triggers it once.
+    // Users of L2BlockStream decide what mode to run it in (_periodically_ vs _manually triggered_).
+    // The default is _manually triggered_.
```
Let's also add docs to `sync` that mention that the process only runs once when `sync` is invoked if `start` was not invoked.

nventuro reviewed (approved):
Lovely

benesjan reviewed (approved):
Feel free to merge this. Feels like we now have consensus on the naming

mverzilli committed:
refactor(PXE): clean up block sync

Block sync is currently handled through collaboration of `Synchronizer`, `SyncDataProvider`, and even `PXEOracleInterface#getAnchorBlock`.

## Current problems
- The names aren't very intuitive (`Synchronizer` and `SyncDataProvider` only handle chain state data at the end of the day).
- Any code running inside `ContractFunctionSimulator` can call `getAnchorBlock` whenever it pleases. This works currently because calls to `Synchronizer` are carefully orchestrated from PXE. But it seems error prone, obscure, and it gives the false impression that function simulation happens on live block data when it actually depends on a static anchor.
- There are multiple ways throughout the app to get the anchor block: `Synchronizer#getSynchedBlock` (unused), `PXEOracleInterface#getAnchorBlock` (unnecessary, inappropriate), `SyncDataProvider#getBlockNumber` (redundant), and `SyncDataProvider#getBlockHeader` (the one that should be the source of truth)
- Many checks are relying on block number when they should rely on block hash.

## Refactor

(note: adjusted naming after discussing with @benesjan)

- Rename `Synchronizer` => `BlockSynchronizer`
- Rename `SyncDataProvider` => `AnchorBlockDataProvider`
- Remove `AnchorBlockDataProvider#getBlockNumber`. `AnchorBlockDataProvider#getBlockHeader` remains the only way to consume current block information for the rest of the app. To get the block number, there's a simple `getBlockNumber` accessor in `BlockHeader`.
- Remove `ExecutionDataProvider#getAnchorBlock`. Instead, explicitly pass anchor block headers as params to functions in "the realm" of `ContractFunctionSimulator` (`run`, `runUtility`, `varifyCurrentClassId`, etc).

## Out of scope of PR

> - Many checks are relying on block number when they should rely on block hash.

This should be tackled on a per use case/feature basis in future PRs, though relying on `getBlockHeader` solely seems aligned in the right direction.

Closes F-225
